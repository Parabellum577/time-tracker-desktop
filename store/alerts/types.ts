import { ActionType } from 'typesafe-actions'

import * as actions from './actions'
import { IError } from '@services/types'

export type AlertsAction = ActionType<typeof actions>

export interface IAlert {
  alertType: 'success' | 'error' | 'warning' | 'coldstart'
  alertMessage: string | IError
}

export interface IAlertsState {
  alertsArray: IAlert[]
}
