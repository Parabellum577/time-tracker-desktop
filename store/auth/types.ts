import { ActionType } from 'typesafe-actions'
import { DeepReadonly } from 'utility-types'

import * as actions from './actions'
import { IError } from '@types'

export type AuthAction = ActionType<typeof actions>

export type IAuthState = DeepReadonly<{
  readonly authorized: boolean
  readonly error?: IError | null
  readonly loading: boolean
  readonly isLoadingData: boolean
}>
