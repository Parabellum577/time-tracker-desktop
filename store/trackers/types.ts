import { ActionType } from 'typesafe-actions'

import { ITrackerWithStatus } from '@types'
import * as actions from './actions'

export type TrackersAction = ActionType<typeof actions>

export interface ITrackerToRender {
  Type: string
  TrackerUrl?: string
  Placeholder?: string
  IsUrlLocked?: boolean
  Name?: string
}

export interface ITrackers {
  userTrackers: ITrackerWithStatus[]
}
