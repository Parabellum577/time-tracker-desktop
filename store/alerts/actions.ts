import { createAsyncAction } from 'typesafe-actions'
import { IError } from '@types'
import { IAlert } from './types'

export const showAlertMessage = createAsyncAction(
  'notifications/SHOW_ALERT_MESSAGE_REQUEST',
  'notifications/SHOW_ALERT_MESSAGE_SUCCESS',
  'notifications/SHOW_ALERT_MESSAGE_FAILURE',
)<IAlert, IAlert, IError>()

export const clearAlertMessage = createAsyncAction(
  'notifications/CLEAR_ALERT_MESSAGE_REQUEST',
  'notifications/CLEAR_ALERT_MESSAGE_SUCCESS',
  'notifications/CLEAR_ALERT_MESSAGE_FAILURE',
)<void, void, IError>()
