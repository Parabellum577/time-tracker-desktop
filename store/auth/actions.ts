import { createAction, createAsyncAction } from 'typesafe-actions'
import { IError } from '@types'
import { ILoginUserPayload, ILoginOTSPayload } from '@services/api-types'

export const loginUser = createAsyncAction('auth/LOGIN_REQUEST', 'auth/LOGIN_SUCCESS', 'auth/LOGIN_FAILURE')<
  ILoginUserPayload,
  void,
  IError
>()

export const loginUserViaOAuth = createAsyncAction(
  'auth/LOGIN_VIA_OAUTH_REQUEST',
  'auth/LOGIN_VIA_OAUTH_SUCCESS',
  'auth/LOGIN_VIA_OAUTH_FAILURE',
)<void, void, IError>()

export const otsLogin = createAsyncAction('auth/OTS_LOGIN_REQUEST', 'auth/OTS_LOGIN_SUCCESS', 'auth/OTS_LOGIN_FAILURE')<
  ILoginOTSPayload,
  ILoginOTSPayload,
  IError
>()

export const loadingDataAfterLogin = createAction('auth/LOADING_DATA_AFTER_LOGIN', resolve => (isLoading: boolean) =>
  resolve(isLoading),
)

export const signOut = createAction('auth/SIGN_OUT', resolve => () => resolve())
