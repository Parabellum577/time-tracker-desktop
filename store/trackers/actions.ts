import { createAsyncAction, createAction } from 'typesafe-actions'
import { ITrackerWithStatus, IError, IUpdateTracker } from '@types'

export const getUserTrackers = createAsyncAction(
  'trackers/GET_USER_TRACKERS_REQUEST',
  'trackers/GET_USER_TRACKERS_SUCCESS',
  'trackers/GET_USER_TRACKERS_FAILURE',
)<void, ITrackerWithStatus[], IError>()

export const removeUserTracker = createAction('trackers/REMOVE_USER_TRACKER', action => (trackerID: number) =>
  action(trackerID),
)

export const addUserTracker = createAction('trackers/ADD_USER_TRACKER', action => (tracker: ITrackerWithStatus) =>
  action(tracker),
)

export const updateUserTracker = createAction('trackers/UPDATE_USER_TRACKER', action => (tracker: IUpdateTracker) =>
  action(tracker),
)

export const clearUserTrackers = createAction('trackers/CLEAR_USER_TRACKERS', action => () => action())
