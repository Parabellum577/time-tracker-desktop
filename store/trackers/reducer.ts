import { getType } from 'typesafe-actions'
import * as _ from 'lodash'

import * as actions from './actions'
import { TrackersAction, ITrackers } from './types'

const initialState: ITrackers = {
  userTrackers: [],
}

export default (state: ITrackers = initialState, action: TrackersAction): ITrackers => {
  switch (action.type) {
    case getType(actions.getUserTrackers.success):
      return { ...state, userTrackers: action.payload }
    case getType(actions.removeUserTracker):
      return {
        ...state,
        userTrackers: [...state.userTrackers.filter(tracker => tracker.ID !== action.payload)],
      }
    case getType(actions.addUserTracker):
      const newTracker = action.payload
      const newStateOfAddedTrackers = [...state.userTrackers, newTracker]
      return { ...state, userTrackers: _.sortBy(newStateOfAddedTrackers, 'ID') }
    case getType(actions.updateUserTracker):
      const trackerToUpdate = state.userTrackers.find(item => item.ID === action.payload.ID)
      const newStateOfUpdatedTrackers = [
        ...state.userTrackers.filter(item => item.ID !== action.payload.ID),
        { ...trackerToUpdate, ...action.payload },
      ]
      return {
        ...state,
        userTrackers: _.sortBy(newStateOfUpdatedTrackers, 'ID'),
      }
    case getType(actions.clearUserTrackers):
      return { ...initialState }
    default:
      return state
  }
}
