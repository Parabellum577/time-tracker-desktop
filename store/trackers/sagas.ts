import { SagaIterator } from 'redux-saga'
import { call, put, take, race } from 'redux-saga/effects'
import { getType } from 'typesafe-actions'

import * as _ from 'lodash'
import * as loginActions from '../auth/actions'
import * as actions from './actions'
import api from '@api'
import { ITrackerWithStatus } from '@types'

export function* getUserTrackersSaga(): SagaIterator {
  while (true) {
    yield race({
      getUserTrackersRequest: take(getType(actions.getUserTrackers.request)),
      userLogin: take(getType(loginActions.loginUser.success)),
    })
    yield call(getUserTrackers)
  }
}

export function* getUserTrackers() {
  try {
    const trackers: ITrackerWithStatus[] = yield call(api.tracker.GetUserTrackers)
    yield put(actions.getUserTrackers.success(trackers))
  } catch (e) {
    yield put(actions.getUserTrackers.failure(e))
  }
}
