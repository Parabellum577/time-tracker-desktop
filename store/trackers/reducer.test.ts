import { ActionType, getType } from 'typesafe-actions'

import * as actions from './actions'
import trackers from './reducer'
import { ITrackers } from './types'

import { SIMPLE_TRACKER } from '../../test-helpers/sample'

describe('[Store] Trackers Reducer', () => {
  it('[Success] [Snapshot] Should handle GetUserTrackers', () => {
    const action: ActionType<typeof actions> = {
      type: getType(actions.getUserTrackers.success),
      payload: [{ ...SIMPLE_TRACKER }],
    }

    const currentState: ITrackers = {
      userTrackers: [{ ...SIMPLE_TRACKER }, { ...SIMPLE_TRACKER }],
    }

    expect(trackers(currentState, action)).toMatchSnapshot()
  })
})
