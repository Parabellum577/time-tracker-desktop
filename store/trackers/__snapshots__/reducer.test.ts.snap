// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Store] Trackers Reducer [Success] [Snapshot] Should handle GetUserTrackers 1`] = `
Object {
  "userTrackers": Array [
    Object {
      "CredentialType": "string",
      "Data": "string",
      "Error": "string",
      "ID": 0,
      "ParentID": 0,
      "Status": true,
      "TrackerUserMail": "string",
      "Type": "string",
      "URL": "string",
    },
  ],
}
`;
